File: src/main/java/pocket/transpiler/js/JavaScriptTranspiler.java
package pocket.transpiler.js;

import org.jetbrains.annotations.NotNull;
import pocket.ast.ModuleFnNameGenerator;
import pocket.ast.Program;
import pocket.transpiler.Transpiler;
import pocket.utility.FileUtility;

import java.nio.file.Path;

public final class JavaScriptTranspiler implements Transpiler {
    @Override
    public @NotNull String transpile(
        @NotNull final String entryFilepath, @NotNull final Program program) {
        final var moduleFnNameGenerator = this.getModuleFnNameGenerator();
        final var directory = Path.of("src/main/resources/pocket/js");
        return new JavaScriptTranspileProgramVisitor(moduleFnNameGenerator, directory)
            .visitProgram(program);
    }

    @Override
    public @NotNull ModuleFnNameGenerator getModuleFnNameGenerator() {
        return (filepath ->
            "$_" + FileUtility.removeExtension(filepath).replace('/', '_').replace('\\', '_'));
    }
}

File: src/main/java/pocket/transpiler/js/JavaScriptTranspiler.kt
package pocket.transpiler.js

import pocket.ast.node.Program
import pocket.transpiler.Transpiler
import java.security.MessageDigest

class JavaScriptTranspiler : Transpiler() {
    /**
     * Maps from absolute paths to function names.
     */
    private val filepathFnNameMap = mutableMapOf<String, String>()

    override fun transpile(program: Program): String {
        return JavaScriptVisitor(this).visitProgram(program)
    }

    override fun fnNameGenerator(filepath: String): String {
        return filepathFnNameMap.getOrPut(filepath) {
            "\$_" + createHashFnName(filepath)
        }
    }

    fun getFnName(targetPath: String): String =
        filepathFnNameMap[targetPath] ?: error("Target path not found: $targetPath")

    private fun createHashFnName(filepath: String): String {
        val fnName = MessageDigest.getInstance("SHA-256")
            .digest(filepath.toByteArray())
            .joinToString("") { "%02x".format(it) }
            .substring(0, 16)

        return if (filepathFnNameMap.containsValue(fnName))
            createHashFnName("$filepath#") else fnName
    }
}